### ISSUE TITLE
Enable/disable context menu items based on item source

### ISSUE THREAD

**Issue by javier-godoy** [vaadin/vaadin-grid-flow#575]
Using Grid ContextMenu there is no way to enable/disable items based on item source.
(Related #490, #439)

**Comment by Duncol** [comment]
Is there a chance for this feature in not so distant future?

**Comment by Duncol** [comment]
Ok, the feature is already on vaadin:master
[(https://github.com/vaadin/vaadin-grid-flow/pull/504/files)]
and I've found out that it is a part of vaadin-grid-flow 4.0 released just today.

**Comment by Springrbua** [comment]
In an application I am currently working on, the items of the `ContextMenu` are different for header, body and footer rows.
I tried to achieve this with the following code:
```
private class Context {
    public Object item;
    public String column;
}

private void configContextMenu(GridContextMenu<Object> contextMenu) {
    Context ctx = new Context();
    GridMenuItem<Object> headerItem = contextMenu.addItem("Header Item");
    GridMenuItem<Object> bodyItem = contextMenu.addItem("Header Item");

    contextMenu.addGridContextMenuOpenedListener(event -> { 
        if (event.isOpened()) {
            context.item = event.getItem().orElse(null);
            context.columnId = event.getColumnId().orElse(null);
            boolean header = context.item == null;
            headerItem .setVisible(header);
            bodyItem .setVisible(!header);
        }
    });
}
```

Most of the times this code works, but there are a few problems with it:
- `event.getItem()` sometimes returns an empty `Optional`, even if the `ContextMenu` has been opened on a body row.
- If an item with a submenu is set to visible `false`, it is hidden as expected, but if you set it's visibility to `true` again, it has lost it's submenu (the arrow is hidden). The menu needs to be reopened to show the submenu.

In my opinion, it should be possible to create the `ContextMenu`, depending on the context in which it has been opened. So basically there should be some kind of `beforeOpenListener` in which I can add / remove items from the `ContextMenu`.

**Comment by pekam** [comment]
> In my opinion, it should be possible to create the ContextMenu, depending on the context in which it has been opened. So basically there should be some kind of beforeOpenListener in which I can add / remove items from the ContextMenu

This kind of callback for before opening the context menu is now implemented as `setDynamicContentHandler`-method:
https://github.com/vaadin/vaadin-grid-flow/pull/748/files#diff-13c58c956b3d2f48298aee2929754b96R226-R247

It also allows to cancel the opening of the menu depending on the target item (or if there's no target item).

Unfortunately the feature is not yet released, but it will most likely be included in a minor release to Vaadin 14.

**Comment by TatuLund** [comment]
Is this solved by https://github.com/vaadin/vaadin-grid-flow/pull/748 ?

**Comment by pekam** [comment]
> Is this solved by #748 ?

It at least provides the API to achieve this. It can be tested with `vaadin-grid-flow 4.1.0.alpha1`.

But I'm not sure if this enable/disable case is properly supported based on this comment:
> If an item with a submenu is set to visible false, it is hidden as expected, but if you set it's visibility to true again, it has lost it's submenu (the arrow is hidden). The menu needs to be reopened to show the submenu.

Well, maybe that problem is worth a separate ticket, if it's still present with the new feature. It's anyway about `setVisible` instead of `setEnabled`. So please create a new ticket for that one if needed. I'm closing this one.
